File: quintus3.5/generic/qplib3.5/library/exit.pl
%   Package: exit
%   Author : Richard A. O'Keefe
%   Updated: 30 Sep 1992
%   Defines: An alternative to halt/0

%   Copyright (C) 1987, Quintus Computer Systems, Inc.  All rights reserved.

:- module(exit, [
	exit/0,
	exit/1
   ]).

sccs_id('"@(#)92/09/30 exit.pl	66.1"').

/*	exit
    and	exit(ExitCode)

    are alternatives to the built-in predicate 'halt'.  My original
    reason for wanting them was to have a way of exiting Prolog
    without getting the closing "Prolog execution halted"banner.
    It then occurred to me that while some UNIX systems provide an
    "onexit" function, some don't, and it would be a good idea to
    have a portable way of doing that.  The best thing would be to
    move this into the supported system, so that QP streams could
    be properly closed by a call to QP_run_exit_forms(x) from user
    C code.  Note that we do NOT redefine the standard C function
    exit(2), though we could.
*/

exit :-
	exit(0).


exit(ExitCode) :-
	integer(ExitCode),
	!,
	(   current_stream(_, _, Stream),
	    close(Stream),
	    fail
	;   flush_output(user_output),
	    flush_output(user_error),
	    'QP_run_exit_forms'(ExitCode)
	;   halt % just in case...
	).
exit(ExitCode) :-
	format(user_error, '~N! Type failure in argument ~w of ~w/~w~n',
	    [1, exit, 1]),
	format(user_error, '! ~w expected, but found ~q~n! Goal: ~q~n',
	    [integer, ExitCode, exit(ExitCode)]),
	exit(1).



foreign_file(library(system(libpl)), ['QP_run_exit_forms']).
foreign('QP_run_exit_forms', 'QP_run_exit_forms'(+integer)).

:- load_foreign_executable(library(system(libpl))).

File: quintus3.5/generic/qplib3.5/library/exit.c
/*  File   : exit.c
    Author : Richard A. O'Keefe
    Updated: 3/25/94
    Defines: QP_run_exit_forms()

    Copyright (C) 1987, Quintus Computer Systems, Inc.  All rights reserved.

    This file exists for two reasons:
    (1) It is awkward to load the system call exit(2) in a portable way
	without going through a file like this.
    (2) This gives us the opportunity to provide an "onexit" feature.

    The OnExitList is a list of function/argument pairs which will be
    called on exit.  You add a pair to this list by calling

	QP_add_exit_form(function, argument);
    or	cleanup = QP_on_exit(function, argument);
  
    The argument type is here assumed to be a char*.
    You remove a pair from the list by calling

	QP_del_exit_form(cleanup, do_it_now);

    If do_it_now is true, the cleanup form is executed.

    So the kind of thing you might do is

	{
	    FILE *f = fopen(....);
	    long cleanup = QP_add_exit_form(fclose, f);

	    ...
	    QP_del_exit_form(cleanup, 1);
	}

    A way of exiting is to call
	QP_run_exit_forms(exit_code);
    which should also close down the QP streams, but currently doesn't.
    {In fact, the closing of QP streams should be done by putting exit
    forms on this list, then they could be closed from C.}
*/

#ifndef	lint
static	char SCCSid[] = "@(#)94/03/25 exit.c	72.1";
#endif/*lint*/

#include <stdlib.h>
#include "quintus.h"            /* [PM] 3.5 QP_malloc */
#include "malloc.h"

typedef struct exit_form_rec *exit_form_ptr;

struct exit_form_rec
    {
	exit_form_ptr	prev;
	exit_form_ptr	succ;
	void	      (*func)(/*char**/);
	char*		arg;
    };

static struct exit_form_rec header =
    {
	/* prev */	(exit_form_ptr)0,
	/* succ */	(exit_form_ptr)0,
	/* func */	exit,
	/* arg  */	(char*)0
    };


exit_form_ptr QP_add_exit_form(func, arg)
    void (*func)();
    char* arg;
    {
	register exit_form_ptr p = Malloc(exit_form_ptr, sizeof *p);

	if (!header.prev) header.prev = header.succ = &header;
	if (p) {
	    p->prev = header.prev;
	    p->succ = &header;
	    p->func = func;
	    p->arg  = arg;
	    p->prev->succ = p;
	    header.prev = p;
	}
	return p;
    }


void QP_del_exit_form(p, do_it_now)
    register exit_form_ptr p;
    int do_it_now;
    {
	if (!p) return;
	if (!p->prev) return;
	p->prev->succ = p->succ;
	p->succ->prev = p->prev;
	p->prev = p->succ = (exit_form_ptr)0;
	if (do_it_now) (p->func)(p->arg);
	Free(p);
    }


void QP_run_exit_forms(exit_code)
    int exit_code;
    {
	register exit_form_ptr p = &header;

	p->arg = (char*)exit_code;
	if (!p->prev) p->prev = p->succ = p;
	for (;; p = p->prev) (p->func)(p->arg);
	/*NOTREACHED*/
    }


